app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+        }
+        try {
+            val succeeded : Boolean = if (DocumentsContract.Document.MIME_TYPE_DIR == mimeType) {
+                newFile.mkdir()
+            } else {
+                newFile.createNewFile()
+            }
+            if (!succeeded) {
+                throw FileNotFoundException("Failed to create document with id " + newFile.path)
+            }
+        } catch (e : IOException) {
+            throw FileNotFoundException("Failed to create document with id " + newFile.path)
+        }
+        return newFile.path
+    }
+
+    override fun deleteDocument(documentId : String?) {
+        val file = getFileForDocId(documentId!!)
+        if (!file.delete()) {
+            throw FileNotFoundException("Couldn't delete document with documentId $documentId")
+        }
+    }
+
+    override fun removeDocument(documentId : String, parentDocumentId : String?) {
+
+        val parent = getFileForDocId(parentDocumentId!!)
+        val file = getFileForDocId(documentId) ?: throw FileNotFoundException("Failed to delete document with id $documentId")
+
+
+        var doesFileParentMatch = false
+        val fileParent = file.parentFile
+        if (fileParent == null || fileParent.equals(parent)) {
+            doesFileParentMatch = true
+        }
+
+        if (parent.equals(file) || doesFileParentMatch) {
+            if (file.delete()) {
+
+            } else {
+                throw FileNotFoundException("Failed to delete document with id $documentId")
+            }
+        } else {
+            throw FileNotFoundException("Failed to delete document with id $documentId")
+        }
+    }
+
+    override fun renameDocument(documentId : String?, displayName : String?) : String? {
+
+        if (displayName == null) {
+            throw FileNotFoundException("Couldn't rename document, new name is null")
+        }
+
+        val sourceFile = getFileForDocId(documentId!!)
+        val sourceParentFile = sourceFile.parentFile ?: throw FileNotFoundException("Failed to rename document. File has no parent.")
+        val destFile = File(sourceParentFile.path, displayName)
+
+        try {
+            val renameSucceeded = sourceFile.renameTo(destFile)
+            if (!renameSucceeded) {
+                throw FileNotFoundException("Failed to rename document. Renamed failed.")
+            }
+        } catch (e : Exception) {
+            throw FileNotFoundException("Couldn't rename document. Error: " + e.message)
+        }
+        return getDocIdForFile(destFile).toString()
+    }
+
+    private fun copyDocument(
+        sourceDocumentId : String, sourceParentDocumentId : String,
+        targetParentDocumentId : String?
+    ) : String? {
+        if (!isChildDocument(sourceParentDocumentId, sourceDocumentId)) {
+            throw FileNotFoundException(
+                "Failed to copy document with id " +
+                        sourceDocumentId + ". Parent is not: " + sourceParentDocumentId
+            )
```

Comment:
`Failed to copy document '$sourceDocumentId' as its parent is not '$sourceParentDocumentId'`